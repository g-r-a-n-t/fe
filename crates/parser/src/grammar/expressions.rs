use crate::ast::{self, CallArg, Expr, Path};
use crate::node::Node;
use crate::{Label, ParseFailed, ParseResult, Parser, Token, TokenKind};

use super::types::parse_generic_args;

use if_chain::if_chain;

// Expressions are parsed in Pratt's top-down operator precedence style.
// See this article for a nice explanation:
// <https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html>

/// Parse an expression, starting with the next token.
pub fn parse_expr(par: &mut Parser) -> ParseResult<Node<Expr>> {
    parse_expr_with_min_bp(par, 0)
}

/// Parse an expression, stopping if/when we reach an operator that binds less
/// tightly than given binding power.
pub fn parse_expr_with_min_bp(par: &mut Parser, min_bp: u8) -> ParseResult<Node<Expr>> {
    let mut expr_head = parse_expr_head(par)?;

    while let Some(op) = par.peek() {
        if let Some(lbp) = postfix_binding_power(op) {
            if lbp < min_bp {
                break;
            }

            expr_head = match op {
                TokenKind::ParenOpen => {
                    let args = parse_call_args(par)?;
                    let span = expr_head.span + args.span;
                    Node::new(
                        Expr::Call {
                            func: Box::new(expr_head),
                            generic_args: None,
                            args,
                        },
                        span,
                    )
                }
                TokenKind::BracketOpen => {
                    par.next()?;
                    let index = parse_expr(par)?;
                    let rbracket = par.expect(
                        TokenKind::BracketClose,
                        "failed to parse subscript expression",
                    )?;
                    let span = expr_head.span + rbracket.span;
                    Node::new(
                        Expr::Subscript {
                            value: Box::new(expr_head),
                            index: Box::new(index),
                        },
                        span,
                    )
                }
                TokenKind::If => {
                    par.next()?;
                    let test = parse_expr(par)?;
                    par.expect(
                        TokenKind::Else,
                        "failed to parse ternary `if-else` expression",
                    )?;
                    let else_val = parse_expr(par)?;
                    let span = expr_head.span + else_val.span;
                    Node::new(
                        Expr::Ternary {
                            if_expr: Box::new(expr_head),
                            test: Box::new(test),
                            else_expr: Box::new(else_val),
                        },
                        span,
                    )
                }
                _ => unreachable!(), // patterns above must match those in `postfix_binding_power`
            };
            continue;
        }

        if matches!(op, TokenKind::Lt) {
            let mut bt_par = par.as_bt_parser();
            if_chain! {
                if let Ok(generic_args) = parse_generic_args(&mut bt_par);
                if matches!(bt_par.peek(), Some(TokenKind::ParenOpen));
                if let Ok(args) = parse_call_args(&mut bt_par);
                then {
                    let span = expr_head.span + args.span;
                    expr_head = Node::new(
                        Expr::Call {
                            func: Box::new(expr_head),
                            generic_args: Some(generic_args),
                            args,
                        },
                        span,
                    );
                    bt_par.accept();
                    continue;
                }
            }
        }

        if let Some((lbp, rbp)) = infix_binding_power(op) {
            if lbp < min_bp {
                break;
            }

            let op_tok = par.next()?;
            let rhs = parse_expr_with_min_bp(par, rbp)?;
            expr_head = infix_op(par, expr_head, &op_tok, rhs)?;
            continue;
        }
        break;
    }

    Ok(expr_head)
}

/// Parse call arguments
pub fn parse_call_args(par: &mut Parser) -> ParseResult<Node<Vec<Node<CallArg>>>> {
    use TokenKind::*;
    let lparen = par.assert(ParenOpen);
    let mut args = vec![];
    let mut span = lparen.span;
    loop {
        if par.peek_or_err()? == ParenClose {
            span += par.next()?.span;
            break;
        }
        let arg = parse_expr(par)?;
        if par.peek_or_err()? == Eq {
            let eq_tok = par.next()?;
            let value = parse_expr(par)?;
            if let Expr::Name(name) = arg.kind {
                let span = arg.span + value.span;
                args.push(Node::new(
                    CallArg {
                        label: Some(Node::new(name, arg.span)),
                        value,
                    },
                    span,
                ));
            } else {
                par.fancy_error(
                    "Syntax error in function call argument list",
                    vec![
                        Label::primary(
                            eq_tok.span,
                            "In a function call, `=` is used for named arguments".to_string(),
                        ),
                        Label::secondary(
                            arg.span,
                            "this should be a function parameter name".to_string(),
                        ),
                    ],
                    vec![],
                );
                return Err(ParseFailed);
            }
        } else {
            let span = arg.span;
            args.push(Node::new(
                CallArg {
                    label: None,
                    value: arg,
                },
                span,
            ));
        }
        if par.peek_or_err()? == Comma {
            par.next()?;
        } else {
            span += par
                .expect(ParenClose, "failed to parse function call argument list")?
                .span;
            break;
        }
    }

    Ok(Node::new(args, span))
}

/// Try to build an expression starting with the given token.
fn parse_expr_head(par: &mut Parser) -> ParseResult<Node<Expr>> {
    use TokenKind::*;

    match par.peek_or_err()? {
        Name | SelfValue | Int | Hex | Octal | Binary | Text | True | False => {
            let tok = par.next()?;
            Ok(atom(par, &tok))
        }
        Plus | Minus | Not | Tilde => {
            let op = par.next()?;
            let operand = parse_expr_with_min_bp(par, prefix_binding_power(op.kind))?;
            Ok(unary_op(&op, operand))
        }
        ParenOpen => parse_group_or_tuple(par),
        BracketOpen => parse_list(par),
        _ => {
            let tok = par.next()?;
            par.unexpected_token_error(
                tok.span,
                format!("Unexpected token while parsing expression: `{}`", tok.text),
                vec![],
            );
            Err(ParseFailed)
        }
    }
}

/// Specifies how tightly a prefix unary operator binds to its operand.
fn prefix_binding_power(op: TokenKind) -> u8 {
    use TokenKind::*;
    match op {
        Not => 65,
        Plus | Minus | Tilde => 135,
        _ => panic!("Unexpected unary op token: {:?}", op),
    }
}

/// Specifies how tightly does an infix operator bind to its left and right
/// operands. See https://docs.python.org/3/reference/expressions.html#operator-precedence
fn infix_binding_power(op: TokenKind) -> Option<(u8, u8)> {
    use TokenKind::*;

    let bp = match op {
        // assignment expr `:=`?
        // lambda?
        // Comma => (40, 41),
        Or => (50, 51),
        And => (60, 61),
        // prefix Not => 65

        // all comparisons are the same
        Lt | LtEq | Gt | GtEq | NotEq | EqEq => (70, 71),

        Pipe => (80, 81),
        Hat => (90, 91),
        Amper => (100, 101),
        LtLt | GtGt => (110, 111),
        Plus | Minus => (120, 121),
        Star | Slash | Percent => (130, 131),
        // Prefix Plus | Minus | Tilde => 135
        StarStar => (141, 140),
        Dot => (150, 151),
        ColonColon => (160, 161),
        _ => return None,
    };
    Some(bp)
}

/// Specifies how tightly a postfix operator binds to its operand.
/// We don't have any "real" postfix operators (like `?` in rust),
/// but we treat `[`, `(`, and ternary `if` as though they're postfix
/// operators.
fn postfix_binding_power(op: TokenKind) -> Option<u8> {
    use TokenKind::*;
    match op {
        If => Some(35), // ternary
        BracketOpen => Some(150),
        ParenOpen => Some(150),
        _ => None,
    }
}

/// Parse a square-bracket list expression, eg. `[1, 2, x]`
fn parse_list(par: &mut Parser) -> ParseResult<Node<Expr>> {
    let lbracket = par.assert(TokenKind::BracketOpen);
    let elts = parse_expr_list(par, TokenKind::BracketClose, None)?;
    let rbracket = par.assert(TokenKind::BracketClose);
    let span = lbracket.span + rbracket.span;
    Ok(Node::new(Expr::List { elts }, span))
}

/// Parse a paren-wrapped expression, which might turn out to be a tuple
/// if it contains commas.
fn parse_group_or_tuple(par: &mut Parser) -> ParseResult<Node<Expr>> {
    use TokenKind::*;
    let lparen = par.assert(ParenOpen);
    if par.peek_or_err()? == ParenClose {
        let rparen = par.next()?;
        let span = lparen.span + rparen.span;
        return Ok(Node::new(Expr::Unit, span));
    }

    let elem = parse_expr(par)?;
    match par.peek_or_err()? {
        ParenClose => {
            // expr wrapped in parens
            let rparen = par.next()?;
            let span = lparen.span + rparen.span;
            Ok(Node::new(elem.kind, span))
        }
        Comma => {
            // tuple
            par.next()?;
            let elts = parse_expr_list(par, ParenClose, Some(elem))?;
            let rparen = par.expect(ParenClose, "failed to parse tuple expression")?;
            let span = lparen.span + rparen.span;
            Ok(Node::new(Expr::Tuple { elts }, span))
        }
        _ => {
            let tok = par.next()?;
            par.unexpected_token_error(
                tok.span,
                "Unexpected token while parsing expression in parentheses",
                vec![],
            );
            Err(ParseFailed)
        }
    }
}

/// Parse some number of comma-separated expressions, until `end_marker` is
/// `peek()`ed.
fn parse_expr_list(
    par: &mut Parser,
    end_marker: TokenKind,
    head: Option<Node<Expr>>,
) -> ParseResult<Vec<Node<Expr>>> {
    let mut elts = vec![];
    if let Some(elem) = head {
        elts.push(elem);
    }
    loop {
        let next = par.peek_or_err()?;
        if next == end_marker {
            break;
        }
        elts.push(parse_expr(par)?);
        match par.peek_or_err()? {
            TokenKind::Comma => {
                par.next()?;
            }
            tk if tk == end_marker => break,
            _ => {
                let tok = par.next()?;
                par.unexpected_token_error(
                    tok.span,
                    "Unexpected token while parsing list of expressions",
                    vec![],
                );
                return Err(ParseFailed);
            }
        }
    }

    Ok(elts)
}

/* node building utils */

/// Create an "atom" expr from the given `Token` (`Name`, `Num`, `Bool`, etc)
fn atom(par: &mut Parser, tok: &Token) -> Node<Expr> {
    use TokenKind::*;

    let expr = match tok.kind {
        Name | SelfValue => Expr::Name(tok.text.to_owned()),
        Int | Hex | Octal | Binary => Expr::Num(tok.text.to_owned()),
        True | False => Expr::Bool(tok.kind == True),
        Text => {
            if let Some(string) = unescape_string(tok.text) {
                Expr::Str(string)
            } else {
                par.error(tok.span, "String contains an invalid escape sequence");
                Expr::Str(tok.text.into())
            }
        }
        _ => panic!("Unexpected atom token: {:?}", tok),
    };
    Node::new(expr, tok.span)
}

fn unescape_string(quoted_string: &str) -> Option<String> {
    let inner = &quoted_string[1..quoted_string.len() - 1];
    unescape::unescape(inner)
}

/// Create an expr from the given infix operator and operands.
fn infix_op(
    par: &mut Parser,
    left: Node<Expr>,
    op: &Token,
    right: Node<Expr>,
) -> ParseResult<Node<Expr>> {
    use TokenKind::*;
    let expr = match op.kind {
        Or | And => bool_op(left, op, right),

        Amper | Hat | Pipe | LtLt | GtGt | Plus | Minus | Star | Slash | Percent | StarStar => {
            bin_op(left, op, right)
        }

        Lt | LtEq | Gt | GtEq | NotEq | EqEq => comp_op(left, op, right),

        Dot => {
            let span = left.span + right.span;
            if let Expr::Name(name) = right.kind {
                Node::new(
                    Expr::Attribute {
                        value: Box::new(left),
                        attr: Node::new(name, right.span),
                    },
                    span,
                )
            } else {
                // TODO: check for float number and say something helpful
                par.fancy_error(
                    "failed to parse attribute expression",
                    vec![Label::primary(right.span, "expected a name")],
                    vec![],
                );
                return Err(ParseFailed);
            }
        }
        ColonColon => {
            let mut path = match left.kind {
                Expr::Name(name) => Path {
                    segments: vec![Node::new(name, left.span)],
                },
                Expr::Path(path) => path,
                _ => {
                    par.fancy_error(
                        "failed to parse path expression",
                        vec![
                            Label::secondary(op.span, "path delimiter".to_string()),
                            Label::primary(left.span, "expected a name"),
                        ],
                        vec![],
                    );
                    return Err(ParseFailed);
                }
            };

            // `right` can't be a Path (rbp > lbp); only valid option is `Name`
            match right.kind {
                Expr::Name(name) => {
                    path.segments.push(Node::new(name, right.span));
                    Node::new(Expr::Path(path), left.span + right.span)
                }
                _ => {
                    par.fancy_error(
                        "failed to parse path expression",
                        vec![
                            Label::secondary(op.span, "path delimiter".to_string()),
                            Label::primary(right.span, "expected a name"),
                        ],
                        vec![],
                    );
                    return Err(ParseFailed);
                }
            }
        }
        _ => panic!("Unexpected infix op token: {:?}", op),
    };
    Ok(expr)
}

/// Create an `Expr::BoolOperation` node for the given operator and operands.
fn bool_op(left: Node<Expr>, op: &Token, right: Node<Expr>) -> Node<Expr> {
    use TokenKind::*;
    let astop = match op.kind {
        And => ast::BoolOperator::And,
        Or => ast::BoolOperator::Or,
        _ => panic!(),
    };

    let span = left.span + right.span;
    Node::new(
        Expr::BoolOperation {
            left: Box::new(left),
            op: Node::new(astop, op.span),
            right: Box::new(right),
        },
        span,
    )
}

/// Create an `Expr::BinOperation` node for the given operator and operands.
fn bin_op(left: Node<Expr>, op: &Token, right: Node<Expr>) -> Node<Expr> {
    use ast::BinOperator::*;
    use TokenKind::*;

    let astop = match op.kind {
        Amper => BitAnd,
        Hat => BitXor,
        Pipe => BitOr,
        LtLt => LShift,
        GtGt => RShift,
        Plus => Add,
        Minus => Sub,
        Star => Mult,
        Slash => Div,
        Percent => Mod,
        StarStar => Pow,
        _ => panic!(),
    };

    let span = left.span + right.span;
    Node::new(
        Expr::BinOperation {
            left: Box::new(left),
            op: Node::new(astop, op.span),
            right: Box::new(right),
        },
        span,
    )
}

/// Create an `Expr::UnaryOperation` node for the given operator and operands.
fn unary_op(op: &Token, operand: Node<Expr>) -> Node<Expr> {
    use ast::UnaryOperator;
    use TokenKind::*;

    let astop = match op.kind {
        Tilde => UnaryOperator::Invert,
        Not => UnaryOperator::Not,
        Minus => UnaryOperator::USub,
        _ => panic!(),
    };

    let span = op.span + operand.span;
    Node::new(
        Expr::UnaryOperation {
            op: Node::new(astop, op.span),
            operand: Box::new(operand),
        },
        span,
    )
}

/// Create an `Expr::CompOperation` node for the given operator and operands.
fn comp_op(left: Node<Expr>, op: &Token, right: Node<Expr>) -> Node<Expr> {
    use ast::CompOperator;
    use TokenKind::*;
    let astop = match op.kind {
        In => todo!("in"), // CompOperator::In,
        Lt => CompOperator::Lt,
        LtEq => CompOperator::LtE,
        Gt => CompOperator::Gt,
        GtEq => CompOperator::GtE,
        NotEq => CompOperator::NotEq,
        EqEq => CompOperator::Eq,
        _ => panic!(),
    };
    let span = left.span + right.span;
    Node::new(
        Expr::CompOperation {
            left: Box::new(left),
            op: Node::new(astop, op.span),
            right: Box::new(right),
        },
        span,
    )
}
