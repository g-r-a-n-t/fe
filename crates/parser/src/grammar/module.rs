use super::contracts::parse_contract_def;
use super::types::{parse_struct_def, parse_type_alias};
use crate::ast::{Module, ModuleStmt, Path, Pragma, Use, UseTree};
use crate::node::{Node, Span};
use crate::{Label, ParseFailed, ParseResult, Parser, TokenKind};

use semver::VersionReq;

/// Parse a [`Module`].
pub fn parse_module(par: &mut Parser) -> ParseResult<Node<Module>> {
    let mut body = vec![];
    loop {
        match par.peek() {
            Some(TokenKind::Newline) => par.expect_newline("module")?,
            Some(TokenKind::Dedent) => {
                par.next()?;
                break;
            }
            None => break,
            Some(_) => {
                let stmt = parse_module_stmt(par)?;
                body.push(stmt);
            }
        }
    }
    let span = Span::zero() + body.first() + body.last();
    Ok(Node::new(Module { body }, span))
}

/// Parse a [`ModuleStmt`].
pub fn parse_module_stmt(par: &mut Parser) -> ParseResult<ModuleStmt> {
    let stmt = match par.peek_or_err()? {
        TokenKind::Pragma => ModuleStmt::Pragma(parse_pragma(par)?),
        TokenKind::Use => ModuleStmt::Use(parse_use(par)?),
        TokenKind::Contract => ModuleStmt::Contract(parse_contract_def(par)?),
        TokenKind::Struct => ModuleStmt::Struct(parse_struct_def(par)?),
        TokenKind::Type => ModuleStmt::TypeAlias(parse_type_alias(par)?),

        // Let these be parse errors for now:
        // TokenKind::Event => todo!("module-level event def"),
        // TokenKind::Name if par.peeked_text() == "from" => parse_from_import(par),
        _ => {
            let tok = par.next()?;
            par.unexpected_token_error(
                tok.span,
                "failed to parse module",
                vec!["Note: expected import, contract, struct, type, or event".into()],
            );
            return Err(ParseFailed);
        }
    };
    Ok(stmt)
}

/// Parse a `use` statement.
/// # Panics
/// Panics if the next token isn't `use`.
pub fn parse_use(par: &mut Parser) -> ParseResult<Node<Use>> {
    let use_tok = par.assert(TokenKind::Use);

    let tree = parse_use_tree(par)?;
    let tree_span = tree.span;

    Ok(Node::new(Use { tree }, use_tok.span + tree_span))
}

/// Parse a `::` delimited path.
pub fn parse_path(par: &mut Parser) -> ParseResult<Node<Path>> {
    let mut names = vec![];

    let name = par.expect_with_notes(TokenKind::Name, "failed to parse path", |_| {
        vec![
            "Note: paths must start with a name".into(),
            "Example: `foo::bar`".into(),
        ]
    })?;

    names.push(Node::new(name.text.to_string(), name.span));

    loop {
        if par.peek() == Some(TokenKind::ColonColon) {
            let delim_tok = par.next()?;

            if par.peek() == Some(TokenKind::Name) {
                let name = par.next()?;

                names.push(Node::new(name.text.to_string(), name.span));
            } else {
                let span =
                    names.first().expect("`names` should not be empty").span + delim_tok.span;

                return Ok(Node::new(
                    Path {
                        names,
                        trailing_delim: true,
                    },
                    span,
                ));
            }
        } else {
            let span = names.first().expect("`names` should not be empty").span
                + names.last().expect("").span;

            return Ok(Node::new(
                Path {
                    names,
                    trailing_delim: false,
                },
                span,
            ));
        }
    }
}

/// Parse a `use` tree.
pub fn parse_use_tree(par: &mut Parser) -> ParseResult<Node<UseTree>> {
    let path = parse_path(par)?;
    let path_span = path.span;

    if path.kind.trailing_delim {
        match par.peek() {
            Some(TokenKind::BraceOpen) => {
                par.next()?;

                let mut children = vec![];
                let close_brace_span;

                loop {
                    children.push(parse_use_tree(par)?);

                    match par.peek() {
                        Some(TokenKind::Comma) => {
                            par.next()?;
                            continue;
                        }
                        Some(TokenKind::BraceClose) => {
                            let tok = par.next()?;
                            close_brace_span = tok.span;
                            break;
                        }
                        _ => {
                            let tok = par.next()?;
                            par.unexpected_token_error(
                                tok.span,
                                "failed to parse `use` tree",
                                vec!["Note: expected a `,` or `}` token".to_string()],
                            );
                            return Err(ParseFailed);
                        }
                    }
                }

                Ok(Node::new(
                    UseTree::Nested {
                        prefix: path,
                        children,
                    },
                    close_brace_span,
                ))
            }
            Some(TokenKind::Star) => {
                par.next()?;
                Ok(Node::new(UseTree::Glob { prefix: path }, path_span))
            }
            _ => {
                let tok = par.next()?;
                par.unexpected_token_error(
                    tok.span,
                    "failed to parse `use` tree",
                    vec!["Note: expected a `*`, `{` or name token".to_string()],
                );
                return Err(ParseFailed);
            }
        }
    } else {
        if par.peek() == Some(TokenKind::As) {
            par.next()?;

            let rename_tok = par.expect(TokenKind::Name, "failed to parse `use` tree")?;
            let rename = Some(Node::new(rename_tok.text.to_string(), rename_tok.span));

            Ok(Node::new(
                UseTree::Simple { path, rename },
                path_span + rename_tok.span,
            ))
        } else {
            Ok(Node::new(UseTree::Simple { path, rename: None }, path_span))
        }
    }
}

/// Parse a `pragma <version-requirement>` statement.
pub fn parse_pragma(par: &mut Parser) -> ParseResult<Node<Pragma>> {
    let tok = par.assert(TokenKind::Pragma);
    assert_eq!(tok.text, "pragma");

    let mut version_string = String::new();
    let mut tokens = vec![];
    loop {
        match par.peek() {
            Some(TokenKind::Newline) => break,
            None => break,
            _ => {
                let tok = par.next()?;
                version_string.push_str(tok.text);
                tokens.push(tok);
            }
        }
    }

    let version_requirement_span = match (tokens.first(), tokens.last()) {
        (Some(first), Some(last)) => first.span + last.span,
        _ => {
            par.error(
                tok.span,
                "failed to parse pragma statement: missing version requirement",
            );
            return Err(ParseFailed);
        }
    };

    match VersionReq::parse(&version_string) {
        Ok(_) => Ok(Node::new(
            Pragma {
                version_requirement: Node::new(version_string, version_requirement_span),
            },
            tok.span + version_requirement_span,
        )),
        Err(err) => {
            par.fancy_error(
                format!("failed to parse pragma statement: {}", err),
                vec![Label::primary(
                    version_requirement_span,
                    "Invalid version requirement",
                )],
                vec!["Example: `^0.5.0`".into()],
            );
            Err(ParseFailed)
        }
    }
}
