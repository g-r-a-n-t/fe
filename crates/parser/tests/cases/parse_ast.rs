use fe_common::diagnostics::print_diagnostics;
use fe_common::utils::ron::to_ron_string_pretty;
use fe_parser::grammar::{contracts, expressions, functions, module, types};
use fe_parser::{ParseResult, Parser};
use insta::assert_snapshot;
use serde::Serialize;
use wasm_bindgen_test::wasm_bindgen_test;

pub fn ast_string<F, T>(test_name: &str, mut parse_fn: F, src: &str) -> String
where
    F: FnMut(&mut Parser) -> ParseResult<T>,
    T: Serialize,
{
    let mut files = fe_common::files::FileStore::new();
    let id = files.add_file(test_name, src);
    let mut parser = Parser::new(src);

    if let Ok(ast) = parse_fn(&mut parser) {
        if !parser.diagnostics.is_empty() {
            print_diagnostics(&parser.diagnostics, id, &files);
            panic!("parse error");
        }
        to_ron_string_pretty(&ast).unwrap()
    } else {
        if parser.diagnostics.is_empty() {
            eprintln!("parsing failed, but no diagnostics were generated. this should be fixed.");
            let next = parser.next();
            if let Ok(tok) = next {
                parser.error(
                    tok.span,
                    "this is the next token at time of parsing failure",
                );
                print_diagnostics(&parser.diagnostics, id, &files);
            } else {
                eprintln!("parser is at end of file");
            }
        } else {
            print_diagnostics(&parser.diagnostics, id, &files);
        }
        panic!("parse failed");
    }
}

macro_rules! test_parse {
    ($name:ident, $parse_fn:expr, $src:expr) => {
        #[test]
        #[wasm_bindgen_test]
        fn $name() {
            if cfg!(target_arch = "wasm32") {
                fe_common::assert_snapshot_wasm!(
                    concat!("snapshots/cases__parse_ast__", stringify!($name), ".snap"),
                    ast_string(stringify!($name), $parse_fn, $src)
                );
            } else {
                assert_snapshot!(ast_string(stringify!($name), $parse_fn, $src));
            }
        }
    };
}

test_parse! { expr_call1, expressions::parse_expr, "foo()" }
test_parse! { expr_call2, expressions::parse_expr, "foo(1,2,x=3)" }
test_parse! { expr_attr1, expressions::parse_expr, "foo.bar[0][y]" }
test_parse! { expr_attr2, expressions::parse_expr, "a[x].b[y](1)" }
test_parse! { expr_num1, expressions::parse_expr, "12345" }
test_parse! { expr_num2, expressions::parse_expr, "00001" }
test_parse! { expr_hex1, expressions::parse_expr, "0xbeefbeef" }
test_parse! { expr_hex2, expressions::parse_expr, "0xFEED1234" }
test_parse! { expr_string, expressions::parse_expr, r#""hi \tmom\n""# }
test_parse! { expr_list, expressions::parse_expr, "[]" }
test_parse! { expr_list2, expressions::parse_expr, "[x, y, z,]" }
test_parse! { expr_ternary, expressions::parse_expr, "x + 1 if y + 2 else z + 3" }
test_parse! { expr_group, expressions::parse_expr, "(1 + 2) * 3" }
test_parse! { expr_tuple1, expressions::parse_expr, "(1,)" }
test_parse! { expr_tuple2, expressions::parse_expr, "(1, 2, \n 3)" }
test_parse! { expr_tuple3, expressions::parse_expr, "(1, (2 + 3), (3 * 4, 5))" }
test_parse! { expr_unit, expressions::parse_expr, "()" }

test_parse! { ops_not, expressions::parse_expr, "x and not y" }
test_parse! { ops_math, expressions::parse_expr, "a + b * -c ** d / e % f" }
test_parse! { ops_neg, expressions::parse_expr, "-x" }
test_parse! { ops_bnot, expressions::parse_expr, "~x" }
// bitwise op precedence: shift > and > xor > or
test_parse! { ops_bit1, expressions::parse_expr, "a & b >> c" }
test_parse! { ops_bit2, expressions::parse_expr, "a ^ b & c" }
test_parse! { ops_bit3, expressions::parse_expr, "a | b ^ c" }
test_parse! { ops_shift, expressions::parse_expr, "a << b >> c" }
test_parse! { ops_bool, expressions::parse_expr, "a or b and c" }

test_parse! { stmt_assert_no_msg, functions::parse_stmt, "assert x == y" }
test_parse! { stmt_assert_msg, functions::parse_stmt, "assert x == y, z" }
test_parse! { stmt_aug_add, functions::parse_stmt, "x += y" }
test_parse! { stmt_aug_sub, functions::parse_stmt, "x -= y" }
test_parse! { stmt_aug_mul, functions::parse_stmt, "x *= y" }
test_parse! { stmt_aug_div, functions::parse_stmt, "x /= y" }
test_parse! { stmt_aug_mod, functions::parse_stmt, "x %= y" }
test_parse! { stmt_aug_and, functions::parse_stmt, "x &= y" }
test_parse! { stmt_aug_or, functions::parse_stmt, "x |= y" }
test_parse! { stmt_aug_xor, functions::parse_stmt, "x ^= y" }
test_parse! { stmt_aug_lsh, functions::parse_stmt, "x <<= y" }
test_parse! { stmt_aug_rsh, functions::parse_stmt, "x >>= y" }
test_parse! { stmt_aug_exp, functions::parse_stmt, "x **= y" }
test_parse! { stmt_emit1, functions::parse_stmt, "emit Foo()" }
test_parse! { stmt_emit2, functions::parse_stmt, "emit Foo(1, 2, x=y)" }
test_parse! { stmt_return1, functions::parse_stmt, "return" }
test_parse! { stmt_return2, functions::parse_stmt, "return x" }
test_parse! { stmt_return3, functions::parse_stmt, "return not x" }
test_parse! { stmt_revert1, functions::parse_stmt, "revert" }

test_parse! { stmt_revert2, functions::parse_stmt, "revert something" }

test_parse! { stmt_if, functions::parse_stmt, "if a:\n b" }
test_parse! { stmt_if2, functions::parse_stmt, "if a:\n b \nelif c:\n d \nelif e: \n f \nelse:\n g" }
test_parse! { stmt_while, functions::parse_stmt, "while a > 5:\n a -= 1" }
test_parse! { stmt_for, functions::parse_stmt, "for a in b[0]:\n pass" }
test_parse! { stmt_var_decl_name, functions::parse_stmt, "let foo: u256 = 1" }
test_parse! { stmt_var_decl_tuple, functions::parse_stmt, "let (foo, bar): (u256, u256) = (10, 10)" }
test_parse! { stmt_var_decl_tuples, functions::parse_stmt, "let (a, (b, (c, d))): x" }

test_parse! { type_def, types::parse_type_alias, "type X = Map<address, u256>" }
test_parse! { type_name, types::parse_type_desc, "MyType" }
test_parse! { type_array, types::parse_type_desc, "address[25]" }
test_parse! { type_3d, types::parse_type_desc, "u256[4][4][4]" }
test_parse! { type_string, types::parse_type_desc, "string<100>" }
test_parse! { type_generic, types::parse_type_desc, "foo<a, b<c>, d[10]>" }
test_parse! { type_generic_int, types::parse_type_desc, "foo<1, 2>" }
test_parse! { type_map1, types::parse_type_desc, "Map<address, u256>" }
test_parse! { type_map2, types::parse_type_desc, "Map<address, Map<u8, u256>>" }
test_parse! { type_map3, types::parse_type_desc, "Map<address, Map<u8, Map<u8, u8>>>" }
test_parse! { type_map4, types::parse_type_desc, "map < address , map < u8, u256 > >" }
test_parse! { type_tuple, types::parse_type_desc, "(u8, u16, address, Map<u8, u8>)" }
test_parse! { type_unit, types::parse_type_desc, "()" }

test_parse! { fn_def, |par| functions::parse_fn_def(par, None), "fn foo21(x: bool, y: address,) -> bool:\n x"}
test_parse! { event_def, types::parse_event_def, "event Foo:\n  x: address\n  idx y: u8" }
test_parse! { empty_event_def, types::parse_event_def, "event Foo:\n  pass" }

test_parse! { pragma1, module::parse_pragma, "pragma 0.1.0" }
test_parse! { pragma2, module::parse_pragma, "pragma 0.1.0-alpha" }
test_parse! { pragma3, module::parse_pragma, "pragma >= 1.2, < 1.5" }

test_parse! { use_simple1, module::parse_use, "use foo::bar" }
test_parse! { use_simple2, module::parse_use, "use foo::bar as baz" }
test_parse! { use_glob, module::parse_use, "use foo::bar::*" }
test_parse! { use_nested1, module::parse_use, "use foo::bar::{bing::*, bang::big, bass as fish, bong::{hello as hi, goodbye}}" }
test_parse! { use_nested2, module::parse_use, r#"use std::bar::{
    bing::*,
    bad::{food as burger, barge::*, bill::bob::{jkl::*}},
    evm as mve
}"#
}
test_parse! { struct_def, types::parse_struct_def, r#"struct S:
  x: address
  pub y: u8
  const z: u8
  pub const a: Map<u8, foo>
"# }
test_parse! { empty_struct_def, types::parse_struct_def, r#"struct S:
  pass
"# }

test_parse! { contract_def, contracts::parse_contract_def, r#"contract Foo:
  x: address
  pub y: u8
  pub const z: Map<u8, address>
  pub fn foo() -> u8:
    return 10
  event Bar:
    idx from: address
"# }

test_parse! { empty_contract_def, contracts::parse_contract_def, r#"contract Foo:
    pass
"# }

test_parse! { module_stmts, module::parse_module, r#"
pragma 0.5.0

use foo::bar::{
    bing as bong,
    food::*
}

type X = Map<u8, u16>

contract A:
    pub const x: u256 = 10

contract B:
    pub x: X
"# }

test_parse! { guest_book, module::parse_module, r#"
type BookMsg = bytes[100]

contract GuestBook:
    pub guest_book: Map<address, BookMsg>

    event Signed:
        idx book_msg: BookMsg

    pub fn sign(self, book_msg: BookMsg):
        self.guest_book[msg.sender] = book_msg

        emit Signed(book_msg=book_msg)

    pub fn get_msg(self, addr: address) -> BookMsg:
        return self.guest_book[addr]
"# }
