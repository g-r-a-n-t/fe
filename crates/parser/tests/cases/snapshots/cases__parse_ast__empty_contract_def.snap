---
source: crates/parser/tests/cases/parse_ast.rs
expression: "ast_string(stringify!(empty_contract_def), contracts::parse_contract_def,\n           r#\"contract Foo:\n    pass\n\"#)"

---
Node(
  kind: Contract(
    name: Node(
      kind: "Foo",
      span: Span(
        start: 9,
        end: 12,
      ),
    ),
    fields: [],
    body: [],
  ),
  span: Span(
    start: 0,
    end: 12,
  ),
)
